hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/MergeIntoHoodieTableCommand.scala

```diff
@@ -554,4 +657,16 @@ object MergeIntoHoodieTableCommand {
     }
   }
 
+  def stripCasting(expr: EqualTo): EqualTo = expr match {
+    case EqualTo(MatchCast(leftExpr, leftTargetType, _, _), MatchCast(rightExpr, rightTargetType, _, _))
+      if leftTargetType.sameType(rightTargetType) => EqualTo(leftExpr, rightExpr)
+    case _ => expr
+  }
+
+  def toStructType(attrs: Seq[Attribute]): StructType =
+    StructType(attrs.map(a => StructField(a.qualifiedName.replace('.', '_'), a.dataType, a.nullable, a.metadata)))
```

Comment:
why do we need to replace perios by underscores?

Reply:
Here we combine the output of joining operation back into schema so that we can assert that the records we receive in `ExpressionPayload` still adhere to the expected schema. Here we need to replace dots, b/c 

 - Names could be qualified (in case of nested structs; note that this schema is essentially a flattened one)
 - Field names can't have dots